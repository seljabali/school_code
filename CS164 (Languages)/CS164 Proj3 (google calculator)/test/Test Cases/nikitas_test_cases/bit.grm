//a bit formula interpreter
%left '+' '-'
%left '>>' '<<'
%left '&' '|'
%right '!' '~'
%%
E -> E '+' E		%{ return n1 + n2 %}
   | E '-' E		%{ return n1 - n2 %}
   | E '&' E 		%{ return n1 & n2 %}
   | E '|' E		%{ return n1 | n2 %}
   | E '>>' E		%{ return n1 >> n2 %}
   | E '<<' E		%{ return n1 << n2 %}
   | '(' E ')'		%{ return n1 %}
   | '!' E		%{ return !n1 %}
   | '~' E		%{ return ~n1 %} //bit inverse
   | '-' E %prec '!'	%{ return -n1 %}
   | NUM;

NUM -> /[1-9][0-9]*/ ;