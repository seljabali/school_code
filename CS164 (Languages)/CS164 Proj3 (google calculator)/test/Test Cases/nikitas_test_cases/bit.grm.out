(Grammar
    (Declarations
        (assoc \+ 0 left)
        (assoc \- 0 left)
        (assoc \>\> 1 left)
        (assoc \<\< 1 left)
        (assoc \& 2 left)
        (assoc \| 2 left)
        (assoc \! 3 right)
        (assoc \~ 3 right)
    )
    (Rules
        (Rule E
            (production [ E, \+, E, ]
                (action { return n1 + n2 })
            )
            (production [ E, \-, E, ]
                (action { return n1 - n2 })
            )
            (production [ E, \&, E, ]
                (action { return n1 & n2 })
            )
            (production [ E, \|, E, ]
                (action { return n1 | n2 })
            )
            (production [ E, \>\>, E, ]
                (action { return n1 >> n2 })
            )
            (production [ E, \<\<, E, ]
                (action { return n1 << n2 })
            )
            (production [ \(, E, \), ]
                (action { return n1 })
            )
            (production [ \!, E, ]
                (action { return !n1 })
            )
            (production [ \~, E, ]
                (action { return ~n1 })
            )
            (production [ \-, E, ]
                (assoc \!)
                (action { return -n1 })
            )
            (production [ NUM, ]
            )
        )
        (Rule NUM
            (production [ [1-9][0-9]*, ]
            )
        )
    )
    (StartSymbol E)
)
